libraries/Microsoft.Bot.Builder.AI.LanguageGeneration/LGActivityGenerator.cs

```diff
@@ -0,0 +1,233 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using AdaptiveCards;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.AI.LanguageGeneration
+{
+    public class LGActivityGenerator
+    {
+        private TemplateEngine templateEngine;
+
+        public const string TextTemplateId = "TextTemplateId";
+
+        public const string AdaptiveCardTemplateId = "AdaptiveCardTemplateId";
+
+        public const string CardTemplateId = "CardTemplateId";
+
+        public const string Attachments = "Attachments";
+
+        public const string Separtor = "Separtor";
+
+        public const string AttachmentLayoutType = "AttachmentLayoutType";
+
+        public LGActivityGenerator(TemplateEngine templateEngine)
+        {
+            this.templateEngine = templateEngine;
+        }
+
+        public Activity Generate(Dictionary<string, object> options, object scope)
+        {
+            /* options schema
+            {
+                "TextTemplateId": "A", // optional if Attachments is not empty.
+                "Separtor": "||", // optional, default "||".
+                "Attachments": [
+                    { "AdaptiveCardTemplateId": "B" },
+                    { "CardTemplateId": "C" },
+                    { "CardTemplateId": "D" },
+                    { "AdaptiveCardTemplateId": "E" }], // optional if TextTemplateId is not null
+                "AttachmentLayoutType": "carousel" // optional, only used for Attachments.Count > 1, default is "carousel", optional value is "list"
+            }
+            */
+            var activity = new Activity();
+            if (options.ContainsKey(TextTemplateId))
+            {
+                var templateId = options[TextTemplateId].ToString();
+                var separtor = options.ContainsKey(Separtor) ? options[Separtor].ToString() : "||";
+                activity = GenerateActivity(templateId, scope, separtor);
```

Comment:
1. Those top level tests (options.ContainsKey) makes your logic  less straghtforward
2. acitity = GenerateActivity, don't seem to align with later activity.Attachments

I would prefer

var activity = new Activity();
(activity.Text, activity.Speak) =  GenTextAndSpeak(options.TextSpeakTemplateId, options.TextSpeakSeperator, data);
activity.Attachments = GenAttachments(options.Attachments, options.Attachment, data)
return activity;

Reply:
fixed